injector/cpu_pressure.go

```diff
@@ -90,83 +75,106 @@ func (i *cpuPressureInjector) Inject() error {
-				i.config.Log.Errorw("failed to join the target cpuset cgroup", "error", err, "core", core, "pid", pid)
+	return nil
+}
 
-				return
-			}
+func (i *cpuPressureInjector) joinCgroup(core int, pid int) bool {
+	// join target CPU cgroup
+	i.config.Log.Infow("joining target CPU cgroup", "core", core, "pid", pid)
 
-			// prioritize the current process
-			i.config.Log.Infow("highering current process priority", "core", core, "pid", pid)
+	if err := i.config.Cgroup.Join("cpu", pid, false); err != nil {
+		i.config.Log.Errorw("failed join the target CPU cgroup", "error", err, "core", core, "pid", pid)
+		return false
+	}
 
-			if err = i.config.ProcessManager.Prioritize(); err != nil {
-				i.config.Log.Errorw("error highering the current process priority", "error", err, "core", core, "pid", pid)
+	// join target cpuset cgroup in case it is used to pin the target on specific cores
+	i.config.Log.Infow("joining target cpuset cgroup", "core", core, "pid", pid)
 
-				return
-			}
+	if err := i.config.Cgroup.Join("cpuset", pid, false); err != nil {
+		i.config.Log.Errorw("failed to join the target cpuset cgroup", "error", err, "core", core, "pid", pid)
+		return false
+	}
 
-			i.config.Log.Infow("starting the stresser", "core", core, "pid", pid)
+	return true
+}
 
-			mutex.Lock()
-			tids = append(tids, pid)
-			i.routines++
-			mutex.Unlock()
+func (i *cpuPressureInjector) startStresser(core int, pid int) {
+	i.config.Log.Infow("starting the stresser", "core", core, "pid", pid)
+	i.config.StresserManager.TrackStresser(core, pid)
+	i.config.Stresser.Stress(i.config.StresserExit)
+}
 
-			wg.Done()
-			i.config.Stresser.Stress(i.config.StresserExit)
-		}(core)
+func (i *cpuPressureInjector) prioritizeStresserProcess(core int, pid int) bool {
+	i.config.Log.Infow("prioritizing the current stresser process", "core", core, "pid", pid)
+
+	if err := i.config.ProcessManager.Prioritize(); err != nil {
+		i.config.Log.Errorw("failed to prioritize the current stresser process", "error", err, "core", core, "pid", pid)
+		return false
 	}
 
-	// wait for stress routines to finish initializing
-	wg.Wait()
+	return true
+}
 
-	if !succeeded {
-		return fmt.Errorf("at least one stresser routine failed to execute")
+func (i *cpuPressureInjector) detectTargetCores() cpuset.CPUSet {
+	if err := i.config.StresserManager.TrackInjectorCores(i.config); err != nil {
+		i.config.Log.Errorw("failed to parse CPUSet", "error", err)
 	}
 
-	i.config.Log.Infow("all routines have been created successfully, now stressing", "routinesPID", tids)
+	cores := i.config.StresserManager.GetCoresToBeStressed()
 
-	return nil
+	i.config.Log.Infow(fmt.Sprintf("target identified to be running on %d cores", cores.Size()), "cores", cores.ToSlice())
+
+	return cores
+}
+
+func (i *cpuPressureInjector) setMaxCPU(cores int) {
+	oldMaxProcs := runtime.GOMAXPROCS(cores)
+	i.config.Log.Infof("changed GOMAXPROCS value from %d to %d", oldMaxProcs, cores)
```

Comment:
I believe this is more the responsibility of the manager to now how to define the MaxCPU.
I would move this method to it.
It would no longer need argument and would use the all cores union to define maxProc (probably protected by a mutex read)